Terminal.Gui/Windows/FileDialog2.cs

```diff
@@ -0,0 +1,2118 @@
+				dlg.UpdateCollectionNavigator ();
+			}
+
+			private static string TrimArrows (string columnName)
+			{
+				return columnName.TrimEnd ('▼', '▲');
+			}
+
+			private static string StripArrows (string columnName)
+			{
+				return columnName.Replace ("▼", string.Empty).Replace ("▲", string.Empty);
+			}
+
+			private void SortColumn (DataColumn clickedCol)
+			{
+				this.GetProposedNewSortOrder (clickedCol, out var isAsc);
+				this.SortColumn (clickedCol, isAsc);
+			}
+
+			private void SortColumn (DataColumn col, bool isAsc)
+			{
+				// set a sort order
+				this.currentSort = col;
+				this.currentSortIsAsc = isAsc;
+
+				this.ApplySort ();
+			}
+
+			private string GetProposedNewSortOrder (DataColumn clickedCol, out bool isAsc)
+			{
+				// work out new sort order
+				if (this.currentSort == clickedCol && this.currentSortIsAsc) {
+					isAsc = false;
+					return $"{clickedCol.ColumnName} DESC";
+				} else {
+					isAsc = true;
+					return $"{clickedCol.ColumnName} ASC";
+				}
+			}
+
+			private void ShowHeaderContextMenu (DataColumn clickedCol, View.MouseEventArgs e)
+			{
+				var sort = this.GetProposedNewSortOrder (clickedCol, out var isAsc);
+
+				var contextMenu = new ContextMenu (
+					e.MouseEvent.X + 1,
+					e.MouseEvent.Y + 1,
+					new MenuBarItem (new MenuItem []
+					{
+						new MenuItem($"Hide {TrimArrows(clickedCol.ColumnName)}", string.Empty, () => this.HideColumn(clickedCol)),
+						new MenuItem($"Sort {StripArrows(sort)}",string.Empty, ()=> this.SortColumn(clickedCol,isAsc)),
+					})
+				);
+
+				contextMenu.Show ();
+			}
+
+			private void HideColumn (DataColumn clickedCol)
+			{
+				var style = this.tableView.Style.GetOrCreateColumnStyle (clickedCol);
+				style.Visible = false;
+				this.tableView.Update ();
+			}
+		}
+
+
+		private class FileDialogRootTreeNode {
+			public FileDialogRootTreeNode (string displayName, DirectoryInfo path)
+			{
+				this.DisplayName = displayName;
+				this.Path = path;
+			}
+
+			public string DisplayName { get; set; }
+			public DirectoryInfo Path { get; set; }
+
+			public override string ToString ()
+			{
+				return this.DisplayName;
+			}
+		}
+
+		private class FileDialogTreeBuilder : ITreeBuilder<object> {
+			public bool SupportsCanExpand => true;
+
+			public bool CanExpand (object toExpand)
+			{
+				return this.TryGetDirectories (NodeToDirectory (toExpand)).Any ();
+			}
+
+			public IEnumerable<object> GetChildren (object forObject)
+			{
+				return this.TryGetDirectories (NodeToDirectory (forObject));
+			}
+
+			internal static DirectoryInfo NodeToDirectory (object toExpand)
+			{
+				return toExpand is FileDialogRootTreeNode f ? f.Path : (DirectoryInfo)toExpand;
+			}
+
+			private IEnumerable<DirectoryInfo> TryGetDirectories (DirectoryInfo directoryInfo)
+			{
+				try {
+					return directoryInfo.EnumerateDirectories ();
+				} catch (Exception) {
+
+					return Enumerable.Empty<DirectoryInfo> ();
+				}
+			}
+
+		}
+	}
+}
```

Comment:
Dunno about you, but I really don't like giant `.cs` files. I'd much rather us adopt using `partial`.